### PR TITLE
feature: implemented keepalive pooling in 'balancer_by_lua*'. (#38)

### PR DESCRIPTION
**Pull request by zhuizhuhaomeng**
Co-authored-by: Thibault Charbonnier <thibaultcha@me.com>

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

File: src/ngx_http_lua_balancer.c [hunk 1]
```diff
@@ -450,64 +793,59 @@ ngx_http_lua_balancer_save_session(ngx_peer_connection_t *pc, void *data)
...
     ngx_url_t              url;
     ngx_http_lua_ctx_t    *ctx;
     ngx_http_upstream_t   *u;
 
-    ngx_http_lua_main_conf_t           *lmcf;
     ngx_http_lua_balancer_peer_data_t  *bp;
 
     if (r == NULL) {
-        *err = "no request found";
+        p = ngx_snprintf(errbuf, *errbuf_size, "no request found");
```
Comment (doujiang24) [inline src/ngx_http_lua_balancer.c:816]: why do we need this change?

File: src/ngx_http_lua_balancer.c [hunk 2]
```diff
@@ -15,10 +15,32 @@
...
+typedef struct {
+    ngx_queue_t                    queue;
+    ngx_queue_t                    hnode;
+    ngx_uint_t                     hash;
+    ngx_http_lua_srv_conf_t       *lscf;
+    ngx_connection_t              *connection;
+    socklen_t                      socklen;
+    ngx_sockaddr_t                 sockaddr;
+    ngx_str_t                      host;
+    u_char                         hostbuf[32];
```
Comment (doujiang24) [inline src/ngx_http_lua_balancer.c:27]: better use a meaning full macro for the hardcode `32`.

And maybe host_data may be a better name? it's not actually a buffer.

Comment (doujiang24) [reply src/ngx_http_lua_balancer.c:27]: And better add a comment for this field.
```
try to avoid allocating memory from the connection pool
```

**Review by doujiang24 — APPROVED** [review]

File: README.markdown [hunk 3]
```diff
@@ -2597,6 +2598,29 @@ This directive was first introduced in the `v0.10.0` release.
...
+keepalive connections to upstream servers that are preserved in the cache of
+each worker process. When this number is exceeded, the least recently used
+connections are closed.
+
+It should be particularly noted that the keepalive directive does not limit the
+total number of connections to upstream servers that an nginx worker process
+can open. The connections parameter should be set to a number small enough to
+let upstream servers process new incoming connections as well.
+
+This directive was first introduced in the `v0.10.19.5` release.
```
Comment (xiaocang) [inline README.markdown:2620]: typo here?
`v0.10.19.5` -> `v0.10.19`

File: doc/HttpLuaModule.wiki [hunk 4]
```diff
@@ -2186,6 +2186,26 @@ When a relative path like <code>foo/bar.lua</code> is given, they will be turned
...
+keepalive connections to upstream servers that are preserved in the cache of
+each worker process. When this number is exceeded, the least recently used
+connections are closed.
+
+It should be particularly noted that the keepalive directive does not limit the
+total number of connections to upstream servers that an nginx worker process
+can open. The connections parameter should be set to a number small enough to
+let upstream servers process new incoming connections as well.
+
+This directive was first introduced in the <code>v0.10.19.5</code> release.
```
Comment (xiaocang) [inline doc/HttpLuaModule.wiki:2207]: Ditto

File: src/ngx_http_lua_balancer.c [hunk 5]
```diff
@@ -520,22 +860,91 @@ ngx_http_lua_ffi_balancer_set_current_peer(ngx_http_request_t *r,
 
     if (ngx_parse_url(r->pool, &url) != NGX_OK) {
         if (url.err) {
-            *err = url.err;
+            p = ngx_snprintf(errbuf, *errbuf_size, "%s", url.err);
```
Comment (xiaocang) [inline src/ngx_http_lua_balancer.c:863]: `errbuf` is undeclared before

**Review by xiaocang — CHANGES_REQUESTED** [review]

File: src/ngx_http_lua_balancer.c [hunk 6]
```diff
@@ -415,12 +574,205 @@ ngx_http_lua_balancer_free_peer(ngx_peer_connection_t *pc, void *data,
...
+    if (c->close || c->read->timedout) {
+        goto close;
+    }
+
+    n = recv(c->fd, buf, 1, MSG_PEEK);
+
+    if (n == -1 && ngx_socket_errno == NGX_EAGAIN) {
+        ev->ready = 0;
+
+        if (ngx_handle_read_event(c->read, 0) == NGX_OK) {
```
Comment (doujiang24) [inline src/ngx_http_lua_balancer.c:759]:
```suggestion
        if (ngx_handle_read_event(c->read, 0) != NGX_OK) {
```

File: src/ngx_http_lua_balancer.c [hunk 7]
```diff
@@ -799,4 +1184,162 @@ ngx_http_lua_ffi_balancer_recreate_request(ngx_http_request_t *r,
...
+    {
+        item = ngx_queue_data(q, ngx_http_lua_balancer_ka_item_t, hnode);
+        if (item->hash != hash) {
+            continue;
+        }
+
+        if (ngx_memn2cmp((u_char *) &item->sockaddr,
+                         (u_char *) sockaddr,
+                         item->socklen, socklen) == 0
+            && name->len == item->host.len
```
Comment (doujiang24) [inline src/ngx_http_lua_balancer.c:1317]: maybe we should move these length compare expressions at the front?
I'm not sure if the C compiler is clever enough to do such optimizations.

File: src/ngx_http_lua_balancer.c [hunk 8]
```diff
@@ -799,4 +1184,162 @@ ngx_http_lua_ffi_balancer_recreate_request(ngx_http_request_t *r,
...
+        name = u->ssl_name;
+    }
+
+    if (name.len == 0) {
+        goto done;
+    }
+
+    /*
+     * ssl name here may contain port, notably if derived from $proxy_host
+     * or $http_host; we have to strip it
```
Comment (doujiang24) [inline src/ngx_http_lua_balancer.c:1244]: better add an example of the `ssl name` value here.

**Comment by Triple-Z** [comment]
Any update on this issue?

**Comment by tzssangglass** [comment]
hi @zhuizhuhaomeng, we found an edge bug in the keepalive pool.

in short, in the balancer phase, when obtaining a connection from the upstream connection pool, the cached attribute of the peer connection is set to 1(`pc->cached = 1`;), indicating that the connection is obtained from the cache.

if an error occurs during the use of this connection, such as "upstream prematurely closed connection" the system will increase the tries attribute of the peer connection by executing `u->peer.tries++`.

tries represents the maximum number of attempts to connect to an upstream server. It equals the normal 1 try + set_more_tries.

the occurrence of `u->peer. tries++` is unexpected and it results in the actual retry count exceeding `set_more_tries`.

here is a reproduce test case:

```perl
# vim:set ft= ts=4 sw=4 et:

use Test::Nginx::Socket::Lua;
use Cwd qw(cwd);

log_level('info');
repeat_each(1);

plan tests => repeat_each() * (blocks() * 6);

my $pwd = cwd();

no_long_string();

check_accum_error_log();
run_tests();

__DATA__

=== TEST 1: sanity
--- http_config
    lua_shared_dict request_counter 1m;
    upstream my_upstream {
        server 127.0.0.1;
        balancer_by_lua_block {
            local balancer = require "ngx.balancer"

            if not ngx.ctx.tries then
                ngx.ctx.tries = 0
            end

            ngx.ctx.tries = ngx.ctx.tries + 1
            ngx.log(ngx.INFO, "tries ", ngx.ctx.tries)

            if ngx.ctx.tries == 1 then
                balancer.set_more_tries(5)
            end
            
            local host = "127.0.0.1"
            local port = 8090;

            local ok, err = balancer.set_current_peer(host, port)
            if not ok then
                ngx.log(ngx.ERR, "failed to set the current peer: ", err)
                return ngx.exit(500)
            end

            balancer.set_timeouts(60000, 60000, 60000)

            local ok, err = balancer.enable_keepalive(60, 100)
            if not ok then
                ngx.log(ngx.ERR, "failed to enable keepalive: ", err)
                return ngx.exit(500)
            end
        }
    }

    server {
        listen 0.0.0.0:8090;
        location /hello {
            content_by_lua_block{                
                local request_counter = ngx.shared.request_counter
                local first_request = request_counter:get("first_request")
                if first_request == nil then
                    request_counter:set("first_request", "yes")
                    ngx.print("hello")
                else
                    ngx.exit(ngx.HTTP_CLOSE)
                end
            }
        }
    }
--- config
    location = /t {
        proxy_pass http://my_upstream;
        proxy_set_header Connection "keep-alive";
        
        rewrite_by_lua_block {
           ngx.req.set_uri("/hello")
        }
    }
--- pipelined_requests eval
["GET /t HTTP/1.1" , "GET /t HTTP/1.1"]
--- response_body eval
["hello", qr/502/]
--- error_code eval
[200, 502]
--- error_log eval
qr/tries 7/
```

in this case:
1. the first request is normal. balancer creates a connection with the upstream and puts the connection into the balancer connection pool after the first request end
2. the second request, pulling a healthy connection from the balancer connection pool and using it, here would mark connection as cached:

	https://github.com/openresty/lua-nginx-module/blob/6738c3a3b0d62b98ff35acf7f7c4a191207da918/src/ngx_http_lua_balancer.c#L1480

 and the upstream shut down the connection by `ngx.exit(ngx.HTTP_CLOSE)`, this is 1 normal try.
3. enter the retry phase, the shutdown just upstream is marked by nginx as `NGX_HTTP_UPSTREAM_FT_ERROR`, and would enter the follow code:
```C
    if (u->peer.cached && ft_type == NGX_HTTP_UPSTREAM_FT_ERROR) {
        /* TODO: inform balancer instead */
        u->peer.tries++;
    }
```
https://github.com/nginx/nginx/blob/e734df6664e70f118ca3140bcef6d4f1750fa8fa/src/http/ngx_http_upstream.c#L4366-L4369

4. `u->peer.tries++` increases the retry times by one.

it seems that Nginx believes that a cached connection improperly closed upstream is not a real retry, so it compensates for a retry.

in the upstream shutdown scenario, this mechanism drains the connections in the balancer connection pool and tries as many times as set by `set_more_tries`.

this results in executing `balancer_by_lua_block` over `set_more_tries`.

**Comment by tzssangglass** [comment]
one way to fix this is to change this mechanism and implement `notify` for `ngx_http_lua_balancer`, like this: https://github.com/Kong/openresty-patches-review/pull/1/files

the another way is not to handle it, we just need to know that it exists, and we can retrieve the `cached` of connection to determine whether the connection was obtained from the balancer connection pool.

**Comment by tzssangglass** [comment]
cc @dndx

**Comment by zhuizhuhaomeng** [comment]
@tzssangglass would you please prepare a patch for this issue?

**Comment by tzssangglass** [comment]
I'd like to know which of the above fixes you prefer

**Comment by zhuizhuhaomeng** [comment]
I prefer the first one(kong's implementation).

**Pull request merged**